Rewrite AMM frontend

rutefig opened the pull request:
# Rewrite AMM Frontend using React

## Overview
This PR implements the AMM (Automated Market Maker) frontend in React, addressing the requirements outlined in [hackathons#5](https://github.com/Hyle-org/hackathons/issues/5). The implementation preserves the core functionality while introducing React best practices, improved component architecture, and performance optimizations.

## Key Changes
- Migrated the frontend architecture to React
- Implemented reusable component patterns
- Optimised data fetching and state management
- Enhanced developer experience with custom hooks and utilities
- Improved type safety with TypeScript integration

## Technical Implementation

### Component Architecture
1. **Tab Components**
   - Represents major functional sections of the AMM
   - Each tab encapsulates its own state and logic
   - Consistent UI patterns across different operations

2. **TokenSelector**
   - Reusable component for token selection
   - Designed for future dynamic token list integration

3. **UI Components**
   - Atomic design pattern implementation
   - Reusable across different tabs
   - Consistent styling and behaviour

### Custom Hooks
1. **useFormSubmission**
   - Standardizes form handling across components
   - Implements validation patterns using Typescript 
   - Reduces boilerplate in form implementations

2. **useHyllar**
   - Centralizes Hylé indexer interactions
   - Maintains independent state per tab while sharing core functionality
   - Reduces redundant API calls
   - Tested with unit tests covering state fetching and response handling

### API Layer
1. **createApiRequest**
   - Factory pattern for API request generation to reduce boilerplate code
   - Unified error handling for Backend and Indexer requests
   - Type-safe request/response handling

### Utilities
1. **camelizeKeys**
   - Ensures TypeScript type consistency
   - Standardizes property naming conventions
   - Improves code maintainability
   - Used for normalizing API responses

## Related Issues
- Closes [hackathons#5](https://github.com/Hyle-org/hackathons/issues/5)

rutefig committed:
basic functionality and interface rewritten using react in a new package

rutefig committed:
all forms submission with hook useSubmissionForm

rutefig committed:
refactor createApiRequest to switch between server or node reqs, get hyllar token state

rutefig committed:
implemented useful getters for useHyllar and docs

rutefig committed:
added feedback for register api responses

rutefig committed:
updating user balance and token supply on components

rutefig committed:
remove console log and minor ux improvement for Faucet component

rutefig committed:
added local and prod env vars

rutefig committed:
updated README

rutefig committed:
simple test for the useHyllar hook

rutefig committed:
removed unnecessary config for now

BertrandD commented:
Hi @rutefig !

Thank you for your contribution!

I had a look at the code and it looks great! Thank you for that!

I tested it locally, and found a few bugs, could you take a look at them please ?

- When registering the user, the payload sent to backend is 
```json
{"username":"bob","password":"test"}
```
where it expected to be
```json
{"username":"bob.hydentity","password":"test"}
```

same for the other pages I guess, the suffix is expected in the different payloads

- On the other pages, the balance is not updated, the yew version polls the node regularly to fetch the settled balance. 
And whatever I write in the "username" field, it balance says:
> Balance: Account .hydentity not found

And I expected

> Balance: Account bob.hydentity not found


- On the approve page, the "spender" should not have a suffix for hydentity. I even suggest you to make this field "disabled" & hard-code "amm" as the value.


If you have any questions, please ask! (here or on telegram, as you prefer)

rutefig committed:
changes from PR review - concat .hydentity to username, default spender to amm

rutefig commented:
Changes made 🫡 
- Concat ".hydentity" to the `username` being sent to the API call
- Fixed the error message for the balance of the not found accounts
- Removed ".hydentity" suffix on spender and disabled the input with the "amm" hardcoded value

Regarding the polling I don't think is necessary to have a polling when is most probably the user will change some times the tab and in this version the state is updated anytime this happens and anytime the token contract selected is changed. But lmk if you prefer the poll I'll implement it.

BertrandD commented on web-react/src/api/endpoints/faucet.ts:
```diff
@@ -0,0 +1,19 @@
+    endpoint: "/faucet",
+    method: "POST",
+    body: {
+      username,
```
```suggestion
      username: username + ".hydentity",
```

BertrandD commented on web-react/src/components/tabs/Swap.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  const { handleSubmit } = useFormSubmission(swap, {
+    onMutate: () => {
+      setMessage("Approving...");
```
```suggestion
      setMessage("Swaping...");
```

BertrandD commented on web-react/src/components/tabs/Swap.tsx:
```diff
@@ -0,0 +1,67 @@
+      setMessage("Approving...");
+    },
+    onError: (error) => {
+      setMessage(`Failed to approve: ${error.message}`);
```
```suggestion
      setMessage(`Failed to swap: ${error.message}`);
```

BertrandD commented on web-react/src/components/tabs/Swap.tsx:
```diff
@@ -0,0 +1,67 @@
+    },
+    onSuccess: () => {
+      setMessage(
+        `Approve successful for user ${username}.hydentity`
```
```suggestion
        `Swap successful for user ${username}.hydentity`
```

BertrandD commented on web-react/src/App.tsx:
```diff
@@ -0,0 +1,53 @@
+const queryClient = new QueryClient();
+
+function App() {
+  const [activeTab, setActiveTab] = useState<TabOption>(TabOption.Approve);
```
I prefer default tab to be the Faucet one :pray: 

```suggestion
  const [activeTab, setActiveTab] = useState<TabOption>(TabOption.Faucet);
```

BertrandD commented:
Looks good, thanks ! Made a few suggestions on minor text changes :)

rutefig committed:
Apply suggestions from code review

Co-authored-by: Bertrand Darbon <bertrand.darbon@gmail.com>

rutefig commented:
> Looks good, thanks ! Made a few suggestions on minor text changes :)

Good catches! Fixed 👍🏼
